lib/ts/recipe/accountlinking/recipe.ts

```diff
@@ -131,184 +134,208 @@ export default class Recipe extends RecipeModule {
     }
 
     // this function returns the user ID for which the session will be created.
-    createPrimaryUserIdOrLinkAccounts = async ({
+    createPrimaryUserIdOrLinkByAccountInfoOrLinkToSessionIfProvided = async ({
         tenantId,
-        user,
+        user: inputUser,
+        recipeUserId,
+        session,
         userContext,
     }: {
         tenantId: string;
         user: User;
+        recipeUserId: RecipeUserId;
+        session: SessionContainerInterface | undefined;
         userContext: UserContext;
-    }): Promise<User> => {
+    }): Promise<
+        | { status: "OK"; user: User }
+        | {
+              status: "LINKING_TO_SESSION_USER_FAILED" | "NON_PRIMARY_SESSION_USER_OTHER_PRIMARY_USER";
+          }
+    > => {
         logDebugMessage("createPrimaryUserIdOrLinkAccounts called");
+        const retry = () => {
+            logDebugMessage("createPrimaryUserIdOrLinkAccounts retrying....");
+            return this.createPrimaryUserIdOrLinkByAccountInfoOrLinkToSessionIfProvided({
+                tenantId,
+                user: inputUser,
+                session,
+                recipeUserId,
+                userContext,
+            });
+        };
+
         // TODO: fix this
-        if (user === undefined) {
+        if (inputUser === undefined) {
             // This can come here if the user is using session + email verification
             // recipe with a user ID that is not known to supertokens. In this case,
             // we do not allow linking for such users.
-            return user;
-        }
-
-        if (user.isPrimaryUser) {
-            return user;
+            return inputUser;
         }
```

Comment:
`inputUser` cannot be undefined according to types so the compiler may think we can never hit this branch. This is a comment/code left over from the initial account linking impl I think. I'll re-check if it can be removed.